### ISSUE TITLE
Question: Global write tokens or recurrent

### ISSUE THREAD

**Issue by IcarusWizard** [lucidrains/recurrent-memory-transformer-pytorch#17]
Hi. Thanks for sharing this work. 

I have a question regarding the write tokens on the input side. In the current version, you always use the global parameter `memory_tokens` on the input side as [here](https://github.com/lucidrains/recurrent-memory-transformer-pytorch/blob/976af1ecba5c1b421a4e204aab31b297c4302b81/recurrent_memory_transformer_pytorch/recurrent_memory_transformer.py#L335C14-L335C14). But in the origin RMT paper, the equation is written as [H^mem H^0 H^mem], and also in Fig. 4 of the new RMDT paper, they draw the arrow from the last write output to both of the read and write input for the next segment. I think they also intend to use the write memory in a recurrent fashion.

**Comment by lucidrains** [comment]
@IcarusWizard ah that's interesting

would be nice if they ablated that

my take is, should make little difference, as in the first layer, the write memories have access to all the read memories

**Comment by IcarusWizard** [comment]
My thought is, if using the recurrent structure, the skip connection in transformer can implicitly make the next memory close to the previous memory.

**Comment by lucidrains** [comment]
@IcarusWizard yup

do you want to see if the latest commit lines up with the paper better?

**Comment by lucidrains** [comment]
@IcarusWizard i'm pretty sure it won't do anything. the way i had it before is akin to an attention pooling step on the first layer, and we know that works well from some other papers

**Comment by IcarusWizard** [comment]
LGTM. There is one small detail that I don't know whether you need stop gradient. In Fig. 2 of the RMT paper, the gradient arrow is pointing to the write, but in Fig. 4 of the RMDT paper, the gradient arrow is point to the read. I think that won't be a big issue as long as the training is stable.

**Comment by lucidrains** [comment]
@IcarusWizard ok, i've made it a hyperparameter

thanks for reporting this!
